track/gpsd.py

```diff
@@ -0,0 +1,185 @@
+    #       and do we even need to do that, or will the gps hw/sw do that for us...?
+    def get_location(self, timeout, need_3d, err_max):
+        if timeout is not None:
+            assert timeout > 0.0
+            t_start = perf_counter()
+
+        if need_3d:
+            fix_ok = lambda fix: fix == GPSFixType.FIX_3D
+        else:
+            fix_ok = lambda fix: fix == GPSFixType.FIX_3D or fix == GPSFixType.FIX_2D
+
+        for report in self.client:
+            if report['class'] != 'TPV': continue
+
+            if 'mode' in report:
+                if report.mode != 0: self.fix_type = GPSFixType(report.mode)
+                else:                self.fix_type = GPSFixType.NO_FIX
+
+            new_location = OrderedDict()
+            new_location['lat'] = report.lat if 'lat' in report else self.INIT_LOC['lat']
+            new_location['lon'] = report.lon if 'lon' in report else self.INIT_LOC['lon']
+            new_location['alt'] = report.alt if 'alt' in report else self.INIT_LOC['alt']
+            self.location = GPSLocation(**new_location) # TODO: make sure this works properly
+
+            new_errors = OrderedDict()
+            new_errors['lat']   = report.epy if 'epy' in report else self.INIT_ERR['lat']
+            new_errors['lon']   = report.epx if 'epx' in report else self.INIT_ERR['lon']
+            new_errors['alt']   = report.epv if 'epv' in report else self.INIT_ERR['alt']
+            new_errors['track'] = report.epd if 'epd' in report else self.INIT_ERR['track']
+            new_errors['speed'] = report.eps if 'eps' in report else self.INIT_ERR['speed']
+            new_errors['climb'] = report.epc if 'epc' in report else self.INIT_ERR['climb']
+            new_errors['time']  = report.ept if 'ept' in report else self.INIT_ERR['time']
+            self.errors = GPSErrors(**new_errors) # TODO: make sure this works properly
+
+            if self._satisfies_criteria(err_max, fix_ok):
+                return self.location
+
+            if timeout is not None:
+                t_now = perf_counter()
+                if t_now - t_start >= timeout:
+                    self._raise_failure(err_max, fix_ok)
+
+    def _satisfies_criteria(self, err_max, fix_ok):
+        if not fix_ok(self.fix_type): return False
+
+        if isnan(self.location.lat): return False
+        if isnan(self.location.lon): return False
+
+        # TODO: make sure this works properly
+        for field in self.errors._fields:
+            if self.errors._asdict()[field] > err_max._asdict()[field]: return False
+
+        return True
+
+    def _raise_failure(self, err_max, fix_ok):
+        reason = self.FailureReason.NONE
+
+        if not fix_ok(self.fix_type): reason |= self.FailureReason.BAD_FIX
+
+        if isnan(self.location.lat): reason |= self.FailureReason.NO_LAT
+        if isnan(self.location.lon): reason |= self.FailureReason.NO_LON
+
+        # TODO: make sure this works properly
```

Comment:
- [ ] Either refactor this so a person of my intelligence can quickly understand it or at least add a descriptive comment that explains what it does

Reply:
Yeah I'll probably add a comment to this awesome construct.